### PR TITLE
[BEAM-2079] Support TextIO as SQL source/sink

### PR DESCRIPTION
**Pull request by xumingming**
Support TextIO as SQL source/sink

### CODE REVIEW

**Comment by xumingming** [comment]
Retest this please.

**Comment by xumingming** [comment]
Retest this please

**Comment by xumingming** [comment]
R: @XuMingmin

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          public void processElement(ProcessContext ctx) {
+            String str = ctx.element();
+
+            StringReader reader = new StringReader(str);
+            CSVRecord rawRecord = null;
+            try {
+              CSVParser parser = csvFormat.parse(reader);
+              rawRecord = parser.getRecords().get(0);
+            } catch (IOException e) {
+              LOG.error("error record: " + str, e);
```
Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java]: throw exception here to report error? same for L76

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        row.addField(idx, Float.valueOf(raw));
+        break;
+      case DOUBLE:
+        row.addField(idx, Double.valueOf(raw));
+        break;
+      case VARCHAR:
+        row.addField(idx, raw);
+        break;
+      default:
+        row.addField(idx, raw);
```
Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java]: throw `BeamSqlUnsupportedException` instead, `row.addField(idx, raw);` always fail here.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOWriter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        BeamSQLRow row = ctx.element();
+        StringWriter writer = new StringWriter();
+
+        try (CSVPrinter printer = csvFormat.print(writer)) {
+          for (int i = 0; i < row.size(); i++) {
+            printer.print(row.getFieldValue(i).toString());
+          }
+          printer.println();
+        } catch (IOException e) {
+          LOG.error("invalid record: " + row, e);
```
Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOWriter.java]: throw exception

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+  @Override
+  public PCollection<BeamSQLRow> expand(PBegin input) {
+    return input.apply("decodeRecord", TextIO.Read.from(filePattern))
+        .apply(ParDo.of(new DoFn<String, BeamSQLRow>() {
+          @ProcessElement
+          public void processElement(ProcessContext ctx) {
+            String str = ctx.element();
+
+            StringReader reader = new StringReader(str);
```
Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java]: need to close `StringReader` at the end of method;

Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java]: missed that :(

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java (hunk 1 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/schema/text/BeamTextCSVTableIOReader.java]: I was thinking that some error records should not fail the whole job. But lets be strict for now.

**Comment by XuMingmin** [comment]
Let's use `TestPipeline` described [here](https://beam.apache.org/contribute/testing/#effective-use-of-the-testpipeline-junit-rule) to update the test cases for `eamTextCSVTableIOWriter.java`, `BeamTextCSVTableIOReader.java`.

**Comment by xumingming** [comment]
Test updated.

**Comment by xumingming** [comment]
R: @jbonofre

**Comment by XuMingmin** [comment]
LGTM
@jbonofre could you also take a look?

**Comment by jbonofre** [comment]
Let me take a look.

**Comment by jbonofre** [comment]
I merged on `DSL_SQL` branch. You can close this PR. Thanks !

**Pull request closed without merging**